source/extensions/filters/http/cache/cache_headers_utils.cc

```diff
@@ -0,0 +1,170 @@
+      is_public_ = true;
+    } else if (directive == "s-maxage") {
+      max_age_ = parseDuration(argument);
+    } else if (!max_age_.has_value() && directive == "max-age") {
+      max_age_ = parseDuration(argument);
+    }
+  }
+}
+
+std::ostream& operator<<(std::ostream& os, const OptionalDuration& duration) {
+  return duration.has_value() ? os << duration.value().count() : os << " ";
+}
+
+std::ostream& operator<<(std::ostream& os, const RequestCacheControl& request_cache_control) {
+  return os << "{"
+            << "must_validate: " << request_cache_control.must_validate_ << ", "
+            << "no_store: " << request_cache_control.no_store_ << ", "
+            << "no_transform: " << request_cache_control.no_transform_ << ", "
+            << "only_if_cached: " << request_cache_control.only_if_cached_ << ", "
+            << "max_age: " << request_cache_control.max_age_ << ", "
+            << "min_fresh: " << request_cache_control.min_fresh_ << ", "
+            << "max_stale: " << request_cache_control.max_stale_ << "}";
+}
+
+std::ostream& operator<<(std::ostream& os, const ResponseCacheControl& response_cache_control) {
+  return os << "{"
+            << "must_validate: " << response_cache_control.must_validate_ << ", "
+            << "no_store: " << response_cache_control.no_store_ << ", "
+            << "no_transform: " << response_cache_control.no_transform_ << ", "
+            << "no_stale: " << response_cache_control.no_stale_ << ", "
+            << "public: " << response_cache_control.is_public_ << ", "
+            << "max_age: " << response_cache_control.max_age_ << "}";
+}
+
+bool operator==(const RequestCacheControl& lhs, const RequestCacheControl& rhs) {
+  return (lhs.must_validate_ == rhs.must_validate_) && (lhs.no_store_ == rhs.no_store_) &&
+         (lhs.no_transform_ == rhs.no_transform_) && (lhs.only_if_cached_ == rhs.only_if_cached_) &&
+         (lhs.max_age_ == rhs.max_age_) && (lhs.min_fresh_ == rhs.min_fresh_) &&
+         (lhs.max_stale_ == rhs.max_stale_);
+}
+
+bool operator==(const ResponseCacheControl& lhs, const ResponseCacheControl& rhs) {
+  return (lhs.must_validate_ == rhs.must_validate_) && (lhs.no_store_ == rhs.no_store_) &&
+         (lhs.no_transform_ == rhs.no_transform_) && (lhs.no_stale_ == rhs.no_stale_) &&
+         (lhs.is_public_ == rhs.is_public_) && (lhs.max_age_ == rhs.max_age_);
+}
+
+SystemTime CacheHeadersUtils::httpTime(const Http::HeaderEntry* header_entry) {
+  if (!header_entry) {
+    return {};
+  }
+  absl::Time time;
+  const std::string input(header_entry->value().getStringView());
+
+  // Acceptable Date/Time Formats per
+  // https://tools.ietf.org/html/rfc7231#section-7.1.1.1
+  //
+  // Sun, 06 Nov 1994 08:49:37 GMT    ; IMF-fixdate
+  // Sunday, 06-Nov-94 08:49:37 GMT   ; obsolete RFC 850 format
+  // Sun Nov  6 08:49:37 1994         ; ANSI C's asctime() format
+  static const auto& rfc7231_date_formats = *new std::array<std::string, 3>{
```

Comment:
use CONSTRUCT_ON_FIRST_USE macro from source/common/common/macros.h.

actually you can do this with a pod:
```
static const char* rfc7231_date_formats[] = {
  "%a, %d %b %Y %H:%M:%S GMT",
  "%A, %d-%b-%y %H:%M:%S GMT",
  "%a %b %e %H:%M:%S %Y"
};
```

Reply:
Actually this function does not belong to the PR it was in the deleted http_cache_utils file:
https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/cache/http_cache_utils.cc

But I'm happy to fix it! Didn't know about that macro, thanks for pointing that out.